pkg/operator/resources/realtimeapi/api.go

```diff
@@ -17,147 +17,157 @@ limitations under the License.
-		isUpdating, err := isAPIUpdating(prevDeployment)
-		if err != nil {
-			return nil, "", err
-		}
-		if isUpdating && !force {
-			return nil, "", ErrorAPIUpdating(api.Name)
+		if err = config.K8s.Update(ctx, &api); err != nil {
+			return nil, "", errors.Wrap(err, "failed to update realtime api resource")
 		}
 
-		if err := config.AWS.UploadJSONToS3(api, config.ClusterConfig.Bucket, api.Key); err != nil {
-			return nil, "", errors.Wrap(err, "upload api spec")
+		if err := config.AWS.UploadJSONToS3(apiSpec, config.ClusterConfig.Bucket, apiSpec.Key); err != nil {
+			return nil, "", errors.Wrap(err, "failed to upload api spec")
 		}
 
-		if err := applyK8sResources(api, prevDeployment, prevService, prevVirtualService); err != nil {
-			return nil, "", err
-		}
-		return api, fmt.Sprintf("updating %s", api.Resource.UserString()), nil
+		return apiSpec, fmt.Sprintf("updating %s", apiConfig.Resource.UserString()), nil
 	}
 
-	// deployment didn't change
-	isUpdating, err := isAPIUpdating(prevDeployment)
-	if err != nil {
-		return nil, "", err
-	}
-	if isUpdating {
-		return api, fmt.Sprintf("%s is already updating", api.Resource.UserString()), nil
-	}
-	return api, fmt.Sprintf("%s is up to date", api.Resource.UserString()), nil
+	return apiSpec, fmt.Sprintf("%s is up to date", apiConfig.Resource.UserString()), nil
 }
 
-func RefreshAPI(apiName string, force bool) (string, error) {
-	prevDeployment, prevService, prevVirtualService, err := getK8sResources(apiName)
-	if err != nil {
-		return "", err
-	} else if prevDeployment == nil || prevVirtualService == nil {
-		return "", errors.ErrorUnexpected("unable to find deployment", apiName)
-	}
+func RefreshAPI(apiName string) (string, error) {
+	ctx := context.Background()
+	var api serverless.RealtimeAPI
+	key := client.ObjectKey{Namespace: consts.DefaultNamespace, Name: apiName}
 
-	isUpdating, err := isAPIUpdating(prevDeployment)
+	err := config.K8s.Get(ctx, key, &api)
 	if err != nil {
-		return "", err
-	}
-
-	if isUpdating && !force {
-		return "", ErrorAPIUpdating(apiName)
+		return "", errors.Wrap(err, "failed to get realtime api resource")
 	}
 
-	apiID, err := k8s.GetLabel(prevDeployment, "apiID")
+	apiSpec, err := operator.DownloadAPISpec(api.Name, api.Annotations["cortex.dev/api-id"])
 	if err != nil {
 		return "", err
 	}
 
-	api, err := operator.DownloadAPISpec(apiName, apiID)
-	if err != nil {
-		return "", err
-	}
+	// generate a new api-id
+	// the deployment-id and spec-id components of the api-id remain unchanged
+	api.Annotations["cortex.dev/api-id"] = ""
+	_, _, _, apiID := api.GetOrCreateAPIIDs()
+	api.Annotations["cortex.dev/api-id"] = apiID
 
-	initialDeploymentTime, err := k8s.ParseInt64Label(prevVirtualService, "initialDeploymentTime")
+	err = config.K8s.Update(ctx, &api)
 	if err != nil {
-		return "", err
+		return "", errors.Wrap(err, "failed to update realtime api resource")
```

Comment:
Why did you change this function? The Refresh operation should not change the api-id. it will change the deployment id and the controller updates the api id

Reply:
The problem is that the refresh was not showing up in `cortex get <api-name>` as it did for the realtime api and as it currently does for the async api. It just does a "ghost" rolling update, which is just one part of the picture. We need the new `api-id` to show up in `cortex get` too.

I set the `deployment-id` instead. Also tested it and it works just like async.

Also, why not set the `api-id` here directly (as long as it's only changing the first component of the ID)? We do this too in `UpdateAPI` so what's the difference?

Reply:
It's in UpdateAPI because it needs to be returned in the response.

For all other cases it should be handled by the controller. The controller should generate a - new api id, so it should (theoretically) work as before. If it is not working it might be a bug that needs to be addressed

Reply:
Right, that makes sense. As of this moment, letting the controller update the `api-id` works. It does coincide with the `api-id` of the s3 file (which we generate in the operator), which is great. Wouldn't have wanted to update the resource and then get it back as that's not efficient, plus we know all other details of the spec.